transports/websys/src/websocket.rs

```diff
@@ -0,0 +1,321 @@
+let transport = websocket::Transport::default()
+    .upgrade(Version::V1)
+    .authenticate(NoiseAuthenticated::xx(&local_key).unwrap())
+    .multiplex(MplexConfig::default())
+    .boxed();
+```
+*/
+#[derive(Default)]
+pub struct Transport;
+
+impl libp2p_core::Transport for Transport {
+    type Output = Connection;
+    type Error = Error;
+    type ListenerUpgrade = Ready<Result<Self::Output, Self::Error>>;
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn listen_on(&mut self, _addr: Multiaddr) -> Result<ListenerId, TransportError<Self::Error>> {
+        Err(TransportError::Other(Error::NotSupported))
+    }
+
+    fn remove_listener(&mut self, _id: ListenerId) -> bool {
+        false
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let ws_url = if let Some(url) = websocket_url(addr) {
+            url
+        } else {
+            return Err(TransportError::Other(Error::NotSupported));
+        };
+
+        Ok(async move {
+            let socket = match WebSocket::new(&ws_url) {
+                Ok(ws) => ws,
+                Err(err) => return Err(Error::JsError(format!("{err:?}"))),
+            };
+
+            Ok(Connection::new(socket))
+        }
+        .boxed())
+    }
+
+    fn dial_as_listener(
+        &mut self,
+        _addr: Multiaddr,
+    ) -> Result<Self::Dial, TransportError<Self::Error>> {
+        Err(TransportError::Other(Error::NotSupported))
+    }
+
+    fn poll(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> std::task::Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>> {
+        Poll::Pending
+    }
+
+    fn address_translation(&self, _listen: &Multiaddr, _observed: &Multiaddr) -> Option<Multiaddr> {
+        None
+    }
+}
+
+// Try to convert Multiaddr to a Websocket url.
+fn websocket_url(addr: Multiaddr) -> Option<String> {
+    let mut protocols = addr.iter();
+    let host_port = match (protocols.next(), protocols.next()) {
+        (Some(Protocol::Ip4(ip)), Some(Protocol::Tcp(port))) => {
+            format!("{ip}:{port}")
+        }
+        (Some(Protocol::Ip6(ip)), Some(Protocol::Tcp(port))) => {
+            format!("[{ip}]:{port}")
+        }
+        (Some(Protocol::Dns(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dns4(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dns6(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dnsaddr(h)), Some(Protocol::Tcp(port))) => {
+            format!("{}:{}", &h, port)
+        }
+        _ => return None,
+    };
+
+    let (scheme, wspath) = match protocols.next() {
+        Some(Protocol::Ws(path)) => ("ws", path.into_owned()),
+        Some(Protocol::Wss(path)) => ("wss", path.into_owned()),
+        _ => return None,
+    };
+
+    // TODO: handle PeerId
+    Some(format!("{scheme}://{host_port}{wspath}"))
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
```

Comment:
Can we make this an opaque error type please?

Reply:
I can do, what can I use as example?

Reply:
Make it a struct with private fields and only a `pub(crate)` constructor.